File: tests/python/test_canonical_scope.py
import pytest
from starling.bus.canonical_scope import (
    CanonicalScopeNull, CanonicalScopeNorm, CanonicalScopeCommitment,
    CanonicalScopeCommonGround, canonical_scope_bytes, scope_of,
)


class _Stmt:
    holder_id = "self"
    subject_kind = "cognizer"
    subject_id = "bob"
    predicate = "p"
    modality = "BELIEVES"
    scope_parties = ()


def test_plain_belief_is_null():
    assert isinstance(scope_of(_Stmt()), CanonicalScopeNull)


def test_commits_is_commitment():
    class S(_Stmt):
        modality = "COMMITS"
        scope_parties = ("bob", "self")
    sc = scope_of(S())
    assert isinstance(sc, CanonicalScopeCommitment)
    assert sc.principal == "self"
    assert sc.beneficiary == "bob"


def test_norm_ought_is_norm():
    class S(_Stmt):
        modality = "NORM_OUGHT"
    sc = scope_of(S())
    assert isinstance(sc, CanonicalScopeNorm)
    assert sc.kind == "obligation"
    assert sc.canonical_bytes() != ""


def test_two_parties_is_common_ground():
    class S(_Stmt):
        scope_parties = ("self", "bob")
    sc = scope_of(S())
    assert isinstance(sc, CanonicalScopeCommonGround)
    assert list(sc.parties_sorted) == ["bob", "self"]


def test_null_canonical_bytes_is_empty():
    assert CanonicalScopeNull().canonical_bytes() == ""
    assert canonical_scope_bytes(CanonicalScopeNull()) == ""


def test_norm_canonical_bytes():
    assert CanonicalScopeNorm(kind="obligation", members_sorted=("a", "b")).canonical_bytes() \
        == "obligation\x1fa\x1fb"


def test_commitment_canonical_bytes():
    assert CanonicalScopeCommitment(principal="alice", beneficiary="bob").canonical_bytes() \
        == "alice\x1fbob"


def test_common_ground_canonical_bytes():
    assert CanonicalScopeCommonGround(parties_sorted=("alice", "bob")).canonical_bytes() \
        == "alice\x1fbob"

File: tests/cpp/test_canonical_scope.cpp
#include "starling/bus/canonical_scope.hpp"
#include "starling/extractor/extracted_statement.hpp"
#include "starling/schema/statement_enums.hpp"
#include <gtest/gtest.h>

using namespace starling;
using namespace starling::bus;

static extractor::ExtractedStatement base() {
    extractor::ExtractedStatement s;
    s.holder_id = "self"; s.subject_kind = "cognizer"; s.subject_id = "bob";
    s.predicate = "p"; s.modality = schema::Modality::BELIEVES;
    return s;
}

TEST(CanonicalScope, PlainBeliefIsNull) {
    auto s = base();
    EXPECT_TRUE(std::holds_alternative<CanonicalScopeNull>(scope_of(s)));
}
TEST(CanonicalScope, CommitsIsCommitment) {
    auto s = base(); s.modality = schema::Modality::COMMITS;
    s.scope_parties = {"bob", "self"};
    auto sc = scope_of(s);
    ASSERT_TRUE(std::holds_alternative<CanonicalScopeCommitment>(sc));
    EXPECT_EQ(std::get<CanonicalScopeCommitment>(sc).principal, "self");
    EXPECT_EQ(std::get<CanonicalScopeCommitment>(sc).beneficiary, "bob");
}
TEST(CanonicalScope, NormOughtIsNorm) {
    auto s = base(); s.modality = schema::Modality::NORM_OUGHT;
    auto sc = scope_of(s);
    ASSERT_TRUE(std::holds_alternative<CanonicalScopeNorm>(sc));
    EXPECT_EQ(std::get<CanonicalScopeNorm>(sc).kind, "obligation");
    EXPECT_FALSE(std::get<CanonicalScopeNorm>(sc).canonical_bytes().empty());
}
TEST(CanonicalScope, NormForbidIsProhibition) {
    auto s = base(); s.modality = schema::Modality::NORM_FORBID;
    auto sc = scope_of(s);
    ASSERT_TRUE(std::holds_alternative<CanonicalScopeNorm>(sc));
    EXPECT_EQ(std::get<CanonicalScopeNorm>(sc).kind, "prohibition");
}
TEST(CanonicalScope, TwoPartiesIsCommonGround) {
    auto s = base(); s.scope_parties = {"self", "bob"};
    auto sc = scope_of(s);
    ASSERT_TRUE(std::holds_alternative<CanonicalScopeCommonGround>(sc));
    EXPECT_EQ(std::get<CanonicalScopeCommonGround>(sc).parties_sorted, (std::vector<std::string>{"bob","self"}));
}
TEST(CanonicalScope, DifferentPartiesDifferentBytes) {
    auto a = base(); a.scope_parties = {"self","bob"};
    auto b = base(); b.scope_parties = {"self","carol"};
    EXPECT_NE(canonical_scope_bytes(scope_of(a)), canonical_scope_bytes(scope_of(b)));
}
